src/sbin/fdisk/fdisk.8
.\"	$MirOS: src/sbin/fdisk/fdisk.8,v 1.8 2008/11/08 23:04:01 tg Exp $
.\"	$OpenBSD: fdisk.8,v 1.51 2006/08/05 14:04:00 ray Exp $
.\"
.\" Copyright (c) 2003, 2004, 2005
.\"	Thorsten "mirabilos" Glaser <tg@mirbsd.org>
.\" Copyright (c) 1997 Tobias Weingartner
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 3, 2002
.Dt FDISK 8
.Os
.Sh NAME
.Nm fdisk
.Nd MBR partition maintenance program
.Sh SYNOPSIS
.Nm fdisk
.Op Fl ieu
.Oo
.Fl c Ar cylinders
.Fl h Ar heads
.Fl s Ar sectors
.Oc
.Op Fl f Ar mbrfile
.Ar device
.Sh DESCRIPTION
On the i386 and other architectures, sector 0 of a bootable hard disc
must contain MBR bootcode, the MBR partition table containing 4 slots,
and a specific magic number (0xAA55).
The 4 slots in the MBR partition table allow a disk drive to be
sub-divided up into chunks known as MBR partitions.
.Pp
On the i386, the BIOS loads sector 0 of the boot disk into memory,
verifies the magic number, and begins executing the MBR bootcode
at the first byte.
The MBR bootcode then searches the MBR partition table for an
.Dq active
MBR partition (indicated by a
.Ql \&*
in the first column), and if one
is found, the boot block from that MBR partition is loaded and executed in
place of the original (MBR) boot block.
.Pp
Some other architectures (like the zaurus), consider sector 0 of the disk
to contain the MBR partition table, but do not use the MBR bootcode at all.
.Sh TYPICAL LAYOUT
When called with no special flags,
.Nm
prints the MBR partition table of the specified device, i.e.,
.Bd -literal -offset 1n
# fdisk wd0
Disk: wd0       geometry: 5168/240/63 [78140160 Sectors]
Offset: 0       Signature: 0xAA55
         Starting       Ending       LBA Info:
 #: id    C   H  S -    C   H  S [       start:      size   ]
------------------------------------------------------------------------
 0: 04    0   1  1 -  170   0 63 [          63:     2570462 ] DOS FAT-16
 1: 00    0   0  0 -    0   0  0 [           0:           0 ] unused
 2: 00    0   0  0 -    0   0  0 [           0:           0 ] unused
*3: A6  170   1  1 - 5167 239 63 [     2570463:    75569697 ] OpenBSD
.Ed
.Pp
It should be noted that the partition table shown above is
invalid and a bad idea in general.
This 78140160 sector (38154MB) disk drive is divided into two MBR
partitions that span the whole disk.
The first MBR partition is a 1225MB FAT-16 DOS partition
(in this case containing a Phoenix BIOS Laptop Suspend-to-Disk file),
the second is a 36GB
.Ox
MBR partition using the remainder of the disk.
The fields of the output are thus:
.Bl -tag -width "start/size"
.It Em "#"
Number identifying each MBR partition table entry.
There are a total of four slots.
.Sq *
denotes the MBR partition which is declared bootable.
.It Em "id"
MBR partition type identifier.
.Mx
reserves the
magic number hexadecimal 27 (39 decimal).
If no MBR partition of type 27 is found,
.Mx
can use an older
.Ox ,
.Nx
or
.Fx
MBR partition (with type 0xA6, 0xA9 and 0xA5, respectively).
.It Em "C/H/S"
These fields provide the starting and ending address of the MBR partition
in BIOS geometry.
.It Em "start/size"
These fields provide the starting sector and size in sectors of the
MBR partition in linear block addresses.
.El
.Pp
.Em NOTE :
The BIOS geometry sectors field is
.Dq 1 based ,
but the start field is
.Dq 0 based .
The CHS values will need to be in the BIOS's geometry
for the system to be able to boot and use the drive correctly.
These values must be kept correctly synchronized or a variety of
problems develop which are very difficult to diagnose.
.Pp
The
.Ox
MBR partition shown above is subdivided further using the
functionality provided by
.Xr disklabel 8 ,
which provides
.Mx
slices.
.Bd -literal -offset 1n
# Inside MBR partition 3: type A6 start 2570463 size 75569697
# /dev/rwd0c:
type: ESDI
disk: ESDI/IDE disk
label: HITACHI_DK13FA-4
flags:
bytes/sector: 512
sectors/track: 63
tracks/cylinder: 16
sectors/cylinder: 1008
cylinders: 16383
total sectors: 78140160
rpm: 3600
interleave: 1
trackskew: 0
cylinderskew: 0
headswitch: 0           # microseconds
track-to-track seek: 0  # microseconds
drivedata: 0

16 partitions:
#           size     offset  fstype [fsize bsize  cpg]
  a:     1740753    2570463  4.2BSD   2048 16384  328 # Cyl  2550*-  4276
  b:      819504    4311216    swap                   # Cyl  4277 -  5089
  c:    78140160          0  unused      0     0      # Cyl     0 - 77519
  d:     2097648    5130720  4.2BSD   2048 16384  328 # Cyl  5090 -  7170
  e:      614880    7228368  4.2BSD   2048 16384  328 # Cyl  7171 -  7780
  g:     2097648    7843248  4.2BSD   2048 16384  328 # Cyl  7781 -  9861
  h:      921312    9940896  4.2BSD   2048 16384  328 # Cyl  9862 - 10775
  i:     2570462         63   MSDOS                   # Cyl     0*-  2549
  j:     3355632   10862208  4.2BSD   2048 16384  328 # Cyl 10776 - 14104
  k:     2097648   14217840  4.2BSD   2048 16384  328 # Cyl 14105 - 16185
  l:     2097648   16315488  4.2BSD   2048 16384  328 # Cyl 16186 - 18266
  m:    59727024   18413136  4.2BSD   2048 16384  328 # Cyl 18267 - 77519
.Ed
.Pp
These
.Mx
filesystems are then mounted as follows using
.Pa /etc/fstab :
.Bd -literal -offset indent
/dev/wd0a / ffs rw,softdep 1 1
/dev/wd0m /home ffs rw,softdep,nodev,nosuid 1 2
/dev/wd0d /tmp ffs rw,softdep,nodev,nosuid 1 2
/dev/wd0g /usr ffs rw,softdep,nodev 1 2
/dev/wd0h /usr/X11R6 ffs rw,softdep,nodev 1 2
/dev/wd0j /usr/local ffs rw,softdep,nodev 1 2
/dev/wd0k /usr/obj ffs rw,softdep,nodev,nosuid 1 2
/dev/wd0l /usr/src ffs rw,softdep,nodev,nosuid 1 2
/dev/wd0e /var ffs rw,softdep,nodev,nosuid 1 2
.Ed
.Sh OPTIONS
The options are as follows:
.Bl -tag -width Ds
.It Xo Fl c Ar cylinders ,
.Fl h Ar heads ,
.Fl s Ar sectors
.Xc
Specifies an alternate BIOS geometry for
.Nm
to use.
By default, an automatic calculation of disk size will be built
using heuristics.
These figures are taken from the in-core disklabel
(see
.Xr disklabel 8 ) ,
or values that
.Em /boot
has passed to the kernel.
.It Fl e
Use the interactive edit of
.Nm
to modify a MBR partition table.
The editor permits configuration of the MBR partition, as well as
extended MBR partitions.
.It Fl f Ar mbrfile
Specifies an alternate MBR template file.
The default file is
.Pa /usr/mdec/mbr .
.It Fl i
Requests that the MBR partition data be re-initialized.
In this mode,
.Nm
will completely overwrite the primary MBR bootcode and MBR partition table
using the default MBR template
.Pa /usr/mdec/mbr
(or the one optionally specified by the
.Fl f
flag).
In the default template, MBR partition number 3 will be configured as an
.Mx
MBR partition spanning from cylinder 0, head 1, sector 1, and extending
to the end of the disk.
This mode is designed to initialize the MBR the very first time.
.It Fl u
Update MBR bootcode, preserving existing MBR partition table.
The MBR bootcode extends from offset 0x000 to the start of the MBR partition table
at offset 0x1BE.
It is similar to the
.Fl i
flag, except the existing MBR partition table is preserved.
This is useful for writing new MBR bootcode onto an existing drive, and is
equivalent to the DOS command
.Dq FDISK /MBR .
Note that this option will overwrite the NT disk signature, if present.
If you would like to preserve the NT disk signature, you can use the
.Em umin
command in the interactive command mode.
Just be damn sure you use a boot code which supports this.
.El
.Sh COMMAND MODE
The
.Fl e
flag causes
.Nm
to enter an interactive command mode.
The prompt contains information about the state of the edit
process.
.Pp
.Dl fdisk:*0>
.Pp
.Sq *
means that the in-memory copy of the boot block has been modified, but
not yet written to disk.
.Pp
0 is the disk offset of the currently selected boot block being edited.
This number could be something other than zero when extended MBR partitions
are being edited (using the
.Em select
subcommand).
.Pp
The list of commands and their explanations are given below.
.Bl -tag -width Ds
.It help
Display a list of commands that
.Nm
understands in the interactive edit mode.
.It manual
Display this manual page.
.It reinit
Initialize the currently selected, in-memory copy of the
boot block.
.It disk
Display the current drive geometry that
.Nm
probed using kernel provided information and various heuristics.
The disk geometry may be changed at this point.
.It edit Ar #
Edit a given table entry in the memory copy of
the current boot block.
Sizes may be adjusted using either in BIOS geometry mode or
sector offsets and sizes.
.It fdef Ar #
Make the given partition table entry bootable.
This does only work with the MirOS boot blocks.
This is not the same as the flag command, which
If an optional value is given, the partition is marked with the given
value, and other partitions are not touched.
\fBAttention:\fR This command should only be used
by people who know what they are doing.
This command can destroy third party bootblocks.
To disable the functionality after it has been enabled,
use the number 9 as parameter.
is commonly used to mark the default entry, too.
.It flag Ar #
Make the given MBR partition table entry bootable.
Only one entry can be marked bootable.
.\" If you wish to boot from an extended
.\" MBR partition, you will need to mark the MBR partition table entry for the
.\" extended MBR partition as bootable.
.\" If an optional value is given, the MBR partition is marked with the given
.\" value, and other MBR partitions are not touched.
If no default entry has been selected with fdef,
or an older, for example from
.Mx 7 ,
or non-MirOS bootblock is used, this entry will be
the default booted entry as well.
.It update
Update the machine MBR bootcode and 0xAA55 signature in the memory copy
of the currently selected boot block.
Note that this option will overwrite an NT disk
signature, if present.
This option will also revert the
.Em fdef
choice to unset, if you use recent
.Mx
boot code.
.It umin
This option performs basically the same operation as the
.Em update
command, but leaving the last part of the machine boot code, which
contains the Microsoft NT disk signature and, shortly before, the
.Em fdef
flag, untouched.
If you do not use recent
.Mx
boot code which supports this, your hard disc will be left
unbootable because this option cuts off the end of the
machine code, possibly containing instructions.
.It select Ar #
Select and load into memory the boot block pointed
to by the extended MBR partition table entry in the current boot block.
.It setpid Ar #
Change the MBR partition
identifier of the given MBR partition table entry.
This command is particularly useful for reassigning
an existing MBR partition to
.Mx .
.It swap Ar # Ar #
Swap two MBR entries.
.It print Op unit
Print the currently selected in-memory copy of the boot
block and its MBR table to the terminal.
If
.Ar unit
is specified as
.Sq b ,
.Sq k ,
.Sq m ,
or
.Sq g ,
the MBR is printed in units of bytes, kibibytes, mebibytes, or gibibytes.
Otherwise the number of sectors is printed.
.It write
Write the in-memory copy of the boot block to disk.
.It exit
Exit the current level of
.Nm fdisk ,
either returning to the
previously selected in-memory copy of a boot block, or exiting the
program if there is none.
.It quit
Exit the current level of
.Nm fdisk ,
either returning to the
previously selected in-memory copy of a boot block, or exiting the
program if there is none.
Unlike
.Em exit
it does write the modified block out.
.It abort
Quit program without saving current changes.
.El
.Sh FILES
.Bl -tag -width /usr/mdec/mbrmgr -compact
.It Pa /usr/mdec/mbrmgr
default MBR template
.El
.Sh SEE ALSO
.Xr fstab 5 ,
.Xr boot_i386 8 ,
.Xr disklabel 8
.Sh CAVEATS
Hand crafted disk layouts are highly error prone.
MBR partitions should start on a cylinder boundary
(head 0, sector 1),
except when starting on track 0,
(these should begin at head 1, sector 1).
MBR partitions should also end at cylinder boundaries.


src/sbin/fdisk/fdisk.c
/*	$OpenBSD: fdisk.c,v 1.43 2006/07/27 04:53:27 ray Exp $	*/

/*
 * Copyright (c) 1997 Tobias Weingartner
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <paths.h>
#include <sys/types.h>
#include <sys/fcntl.h>
#include <sys/disklabel.h>
#include "disk.h"
#include "user.h"

#define _PATH_MBR _PATH_BOOTDIR "mbrmgr"
static unsigned char builtin_mbr[] = {
#include "mbrcode.h"
};

__RCSID("$MirOS: src/sbin/fdisk/fdisk.c,v 1.4 2006/09/20 20:03:30 tg Exp $");

static __dead void
usage(void)
{
	extern char * __progname;

	fprintf(stderr, "usage: %s "
	    "[-ieu] [-c cylinders -h heads -s sectors] [-f mbrfile] device\n"
	    "\t-i: initialise disk with virgin MBR\n"
	    "\t-u: update MBR code, preserve partition table\n"
	    "\t-e: edit MBRs on disk interactively\n"
	    "\t-f: specify non-standard MBR template\n"
	    "\t-chs: specify disk geometry\n"
	    "`disk' may be of the forms: sd0 or /dev/rsd0c.\n",
	    __progname);
	exit(1);
}


int
main(int argc, char *argv[])
{
	int ch, fd;
	int i_flag = 0, m_flag = 0, u_flag = 0;
	int c_arg = 0, h_arg = 0, s_arg = 0;
	disk_t disk;
	DISK_metrics *usermetrics;
#if defined(__amd64__) || defined(__i386__) || defined (__powerpc__)
	const char *mbrfile = _PATH_MBR;
#else
	const char *mbrfile = NULL;
#endif
	mbr_t mbr;
	char mbr_buf[DEV_BSIZE];

	while ((ch = getopt(argc, argv, "ieuf:c:h:s:")) != -1) {
		const char *errstr;

		switch(ch) {
		case 'i':
			i_flag = 1;
			break;
		case 'u':
			u_flag = 1;
			break;
		case 'e':
			m_flag = 1;
			break;
		case 'f':
			mbrfile = optarg;
			break;
		case 'c':
			c_arg = strtonum(optarg, 1, 2097151, &errstr);
			if (errstr)
				errx(1, "Cylinder argument %s [1..2097151].",
				    errstr);
			break;
		case 'h':
			h_arg = strtonum(optarg, 1, 256, &errstr);
			if (errstr)
				errx(1, "Head argument %s [1..256].", errstr);
			break;
		case 's':
			s_arg = strtonum(optarg, 1, 63, &errstr);
			if (errstr)
				errx(1, "Sector argument %s [1..63].", errstr);
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	/* Argument checking */
	if (argc != 1)
		usage();
	else
		disk.name = argv[0];

	/* Put in supplied geometry if there */
	if (c_arg | h_arg | s_arg) {
		usermetrics = malloc(sizeof(DISK_metrics));
		if (usermetrics != NULL) {
			if (c_arg && h_arg && s_arg) {
				usermetrics->cylinders = c_arg;
				usermetrics->heads = h_arg;
				usermetrics->sectors = s_arg;
				usermetrics->size = c_arg * h_arg * s_arg;
			} else
				errx(1, "Please specify a full geometry with [-chs].");
		}
	} else
		usermetrics = NULL;

	/* Get the geometry */
	disk.real = NULL;
	if (DISK_getmetrics(&disk, usermetrics))
		errx(1, "Can't get disk geometry, please use [-chs] to specify.");


	/* Print out current MBRs on disk */
	if ((i_flag + u_flag + m_flag) == 0)
		exit(USER_print_disk(&disk));

	/* Parse mbr template, to pass on later */
	if (mbrfile != NULL && (fd = open(mbrfile, O_RDONLY)) == -1) {
		warn("%s", mbrfile);
		warnx("using builtin MBR");
		mbrfile = NULL;
	}
	if (mbrfile == NULL) {
		memcpy(mbr_buf, builtin_mbr, sizeof(mbr_buf));
	} else {
		MBR_read(fd, 0, mbr_buf);
		close(fd);
	}
	MBR_parse(&disk, mbr_buf, 0, 0, &mbr);

	/* Now do what we are supposed to */
	if (i_flag || u_flag)
		if (USER_init(&disk, &mbr, u_flag) == -1)
			err(1, "error initializing MBR");

	if (m_flag)
		USER_modify(&disk, &mbr, 0, 0);

	return (0);
}
